toolchain/check/impl_validation.cpp

```diff
@@ -0,0 +1,456 @@
+    -> bool {
+  if (impl_a.is_local && impl_b.is_local) {
+    CARBON_DIAGNOSTIC(FinalImplOverlapsSameFile, Error,
+                      "`final impl` overlaps with `final impl` from same file "
+                      "outside a `match_first` block");
+    auto builder = context.emitter().Build(impl_b.latest_decl_id,
+                                           FinalImplOverlapsSameFile);
+    CARBON_DIAGNOSTIC(FinalImplOverlapsSameFileNote, Note,
+                      "other `final impl` here");
+    builder.Note(impl_a.latest_decl_id, FinalImplOverlapsSameFileNote);
+    builder.Emit();
+    return true;
+  }
+
+  return false;
+}
+
+static auto ValidateImplsForInterface(
+    Context& context, llvm::ArrayRef<ImplAndInterface> impls_and_interface)
+    -> void {
+  // Range over `SemIR::ImplId` only. It'd be nice to make this the function
+  // parameter but we don't have a concept to express that outside the (banned)
+  // std::ranges.
+  auto impl_ids = llvm::map_range(impls_and_interface,
+                                  [=](ImplAndInterface impl_and_interface) {
+                                    return impl_and_interface.impl_id;
+                                  });
+
+  // All `impl`s we look at here have the same `InterfaceId` (though different
+  // `SpecificInterface`s in their specifics. So we can grab the ImportIRId for
+  // the interface a single time up front.
+  auto interface_decl_id =
+      context.interfaces()
+          .Get(impls_and_interface[0].interface.interface_id)
+          .first_owning_decl_id;
+  auto interface_ir_id = GetIRId(context, interface_decl_id);
+
+  // TODO: We should revisit this and look for a way to do these checks in less
+  // than quadratic time. From @zygoloid: Possibly by converting the set of
+  // impls into a decision tree.
+  for (auto [index_a, impl_a_id] : llvm::enumerate(impl_ids)) {
+    auto impl_a = GetImplInfo(context, impl_a_id);
+
+    if (impl_a.is_final && impl_a.is_local) {
+      // =======================================================================
+      /// Rules for an individual final impl and its internal values.
+      // =======================================================================
+      if (DiagnoseFinalImplNotInSameFileAsRootSelfTypeOrInterface(
+              context, impl_a, interface_ir_id)) {
+        continue;
+      }
+    }
+
+    for (auto impl_b_id : llvm::drop_begin(impl_ids, index_a + 1)) {
+      auto impl_b = GetImplInfo(context, impl_b_id);
+
+      if (!impl_a.is_final && !impl_b.is_final) {
+        // =====================================================================
+        // Rules between two non-final impls.
+        //
+        // It's possible to write invalid non-final impls in two files that do
+        // not know about each other, and they are only invalid when they are
+        // both imported and visible together. So we need to compare non-local
+        // impls with each other here.
```

Comment:
Once the orphan rule is implemented, this should not be possible across files, so this should only be done if they are both local.